### PR TITLE
ggml : move CPU backend to a separate file

### PR DESCRIPTION
**Pull request by slaren**
Moves the ggml code specific to the CPU backend to a separate file.

This is an initial step to separate the core ggml library from the CPU backend. In the future, this will allow:
- Building other backends as a shared library, without having to link them to the CPU backend
- Building the core ggml library with only the base instruction set for the ABI, and load an optimized version of the CPU backend dynamically

Additionally:
- Removes the optimization interface, since it has dependencies to the CPU backend, and would be removed in https://github.com/ggerganov/ggml/pull/988 regardless
- Removes the baby-llama example since it depends on the opt interface

### CODE REVIEW

**Review by JohannesGaessler — COMMENTED** [review]
Are there also plans to split `ggml-cpu.c` into multiple smaller files like was done for CUDA?

(I did not really look at `ggml.c` and `ggml-cpu.c` since I think it's not feasible.)

File: ggml/include/ggml.h [hunk 1]
```diff
@@ -797,8 +748,7 @@ extern "C" {
             int64_t ne2,
             int64_t ne3);
 
-    GGML_API struct ggml_tensor * ggml_new_i32(struct ggml_context * ctx, int32_t value);
-    GGML_API struct ggml_tensor * ggml_new_f32(struct ggml_context * ctx, float value);
+    GGML_API void * ggml_new_buffer(struct ggml_context * ctx, size_t nbytes);
```
[the remark this answers is not in the document]
Comment (JohannesGaessler) [reply ggml/include/ggml.h:751]: From what I can tell this function is used to allocate `work_data` in `ggml-cpu.c` so I am assuming this function was simply added due to `ggml_context` being only visible in `ggml.c`.

[the remark this answers is not in the document]
Comment (slaren) [reply ggml/include/ggml.h:751]: Yes, this is correct. It is used in `ggml_graph_compute_with_ctx`.

File: ggml/src/ggml-rpc.cpp [hunk 2]
```diff
@@ -1328,7 +1321,7 @@ static const struct ggml_backend_device_i ggml_backend_rpc_device_i = {
     /* .init_backend         = */ ggml_backend_rpc_device_init,
     /* .get_buffer_type      = */ ggml_backend_rpc_device_get_buffer_type,
     /* .get_host_buffer_type = */ NULL,
-    /* .buffer_from_host_ptr = */ ggml_backend_rpc_device_buffer_from_ptr,
+    /* .buffer_from_host_ptr = */ NULL,
```
[the remark this answers is not in the document]
Comment (slaren) [reply ggml/src/ggml-rpc.cpp:1324]: It is a minor bug fix. `ggml_backend_rpc_device_get_props` was correct in reporting that `buffer_from_host_ptr` is not supported, so this function was never called, but nonetheless it shouldn't be there.

**Comment by slaren** [comment]
> Are there also plans to split `ggml-cpu.c` into multiple smaller files like was done for CUDA?

Yes, I think that would be great. We should also adapt it to C++ and use templates to avoid duplicating the code of the operations for each type.

**Review by JohannesGaessler — APPROVED** [review]

**Comment by ggerganov** [comment]
Looking into this now.

**Comment by ggerganov** [comment]
Isn't this going to produce thread sanitizer data race warnings on the `is_first_call` var?

https://github.com/ggerganov/llama.cpp/blob/bf95fffc6fa7a257c43aeb7b6ff47d78af9c9225/ggml/src/ggml.c#L1424-L1443

**Review by ggerganov — APPROVED** [review]

**Comment by chaxu01** [comment]
@slaren this commit 9f40989 breaks q4_0_4_8 on Arm CPU
```
./bin/llama-cli -m llama-2-7b-chat.Q4_0_4_8.gguf -p "Write a code in C for bubble sorting" -n 32 -t 4 -ngl 0
...
Assertion failed: (!isnan(wp[i])Assertion failed: (!isnan(wp[i])Assertion failed: (!isnan(wp[i])Assertion failed: (!isnan(wp[i])), function ggml_compute_forward), function ggml_compute_forward), function ggml_compute_forward_soft_max_f32, file ggml-cpu.c, _soft_max_f32, file ggml-cpu.c, _soft_max_f32, file ggml-cpu.c, ), function ggml_compute_forwardline 8904.
```
It works OK for commit 08828a6.

**Comment by QingtaoLi1** [comment]
@slaren Why the `is_first_call` in `ggml_init()` in ggml.c is set reversely, i.e. is false when first called, while true later?

**Pull request merged**